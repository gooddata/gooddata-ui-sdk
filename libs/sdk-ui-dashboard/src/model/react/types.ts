// (C) 2021 GoodData Corporation
import { IAnalyticalBackend, IDashboard, IWorkspacePermissions } from "@gooddata/sdk-backend-spi";
import { ObjRef } from "@gooddata/sdk-model";
import { DashboardEventHandler } from "../eventHandlers/eventHandler";
import { DashboardDispatch, DashboardState } from "../store";
import { DashboardConfig, DashboardModelCustomizationFns } from "../types/commonTypes";
import React from "react";
import { ReactReduxContextValue } from "react-redux";

/**
 * Subset of IDashboardProps required during initialization of the dashboard component's store.
 *
 * @internal
 */
export interface IDashboardStoreProviderProps {
    backend?: IAnalyticalBackend;
    workspace?: string;
    dashboard?: ObjRef | IDashboard;
    filterContextRef?: ObjRef;
    eventHandlers?: DashboardEventHandler[];
    config?: DashboardConfig;
    permissions?: IWorkspacePermissions;
    onStateChange?: (state: DashboardState, dispatch: DashboardDispatch) => void;
    onEventingInitialized?: (
        registerEventHandler: (handler: DashboardEventHandler) => void,
        unregisterEventHandler: (handler: DashboardEventHandler) => void,
    ) => void;
    additionalReduxContext?: React.Context<ReactReduxContextValue>;
    customizationFns?: DashboardModelCustomizationFns;
}
