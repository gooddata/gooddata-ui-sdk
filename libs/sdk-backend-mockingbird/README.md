# Static / Mock implementation of Analytical Backend SPI

[![npm version](https://img.shields.io/npm/v/@gooddata/sdk-backend-mockingbird)](https://www.npmjs.com/@gooddata/sdk-backend-mockingbird)&nbsp;
[![npm monthly downloads](https://img.shields.io/npm/dm/@gooddata/sdk-backend-mockingbird)](https://npmcharts.com/compare/@gooddata/sdk-backend-mockingbird?minimal=true)&nbsp;
![typescript](https://img.shields.io/badge/typescript-first-blue?logo=typescript)

This package is a part of the [GoodData.UI SDK](https://sdk.gooddata.com/gooddata-ui/docs/about_gooddataui.html).
To learn more, check [the source monorepo](https://github.com/gooddata/gooddata-ui-sdk).

This package contains test support and mock implementations of various concepts in the [Analytical Backend](https://www.npmjs.com/package/@gooddata/sdk-backend-spi).

## Dummy Backend

This implementation of Analytical Backend SPI focuses on the execution branch of the SPI; it creates
and sets up Prepared Execution just like any other implementation would. When the prepared execution
is started (execute()) it either (based on configuration):

-   Returns an empty result which only contains execution definition and has the result dimensions empty.
    Reading any data views from this result will return empty data views.

-   Returns an empty result which only contains execution definition and has the result dimensions empty.
    Reading any data views from this result will yield NoDataError

    (Note: this is closer to how real backend implementations behave)

Purpose:

-   use in unit tests which need to verify whether the execution is prepared correctly
    for instance functions which transform props => prepared execution

-   use in unit tests which need to verify whether code works correctly with execution definition
    stored in result / data view / data view facade

-   use in component-level 'smoke tests' (e.g. something renders or happens, we don't care about the details)

## Legacy Recorded Backend

This implementation of Analytical Backend SPI for serves results recorded on the disk. It is called
legacy, because it works with recordings that were created pre-8.0, using an unknown method and are
tightly coupled with Execute AFM types themselves.

This legacy implementation was essential during the initial refactoring of the SDK to quickly get the
existing tests green. However, it should not be used in any new tests and should not be enhanced any further.

Instead, our efforts need to be focused on enhancing the non-legacy recorded backend and having all
tests (eventually also old tests) run against the reference workspace.

## Composite Backend

This implementation of Analytical Backend SPI allows merging multiple analytical backend instances into one. Each
component declares what for what workspace it has the data and the composite delegates all requests for that
workspace into the respective component.

## License

(C) 2017-2022 GoodData Corporation

This project is under MIT License. See [LICENSE](https://github.com/gooddata/gooddata-ui-sdk/blob/master/libs/sdk-backend-mockingbird/LICENSE).
