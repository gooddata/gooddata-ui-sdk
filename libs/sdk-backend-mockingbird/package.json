{
    "name": "@gooddata/sdk-backend-mockingbird",
    "version": "8.6.0-alpha.41",
    "author": "GoodData",
    "description": "Mock GoodData Backend SPI implementation",
    "repository": {
        "type": "git",
        "url": "https://github.com/gooddata/gooddata-ui-sdk.git",
        "directory": "libs/sdk-backend-mockingbird"
    },
    "license": "MIT",
    "main": "dist/index.js",
    "module": "esm/index.js",
    "browser": "dist/index.js",
    "typings": "esm/index.d.ts",
    "sideEffects": false,
    "files": [
        "dist/**/*.js",
        "dist/**/*.json",
        "dist/**/*.d.ts",
        "dist/**/*.map",
        "esm/**/*.js",
        "esm/**/*.json",
        "esm/**/*.d.ts",
        "esm/**/*.map"
    ],
    "config": {
        "eslint": "-c .eslintrc.js --ext ts src/"
    },
    "scripts": {
        "clean": "rm -rf ci dist esm coverage *.log && jest --clearCache",
        "dev": "tsc -p tsconfig.dev.json --watch",
        "build": "concurrently \"npm run build-cjs\" \"npm run build-esm\" && npm run api-extractor",
        "build-cjs": "tsc -p tsconfig.build.json",
        "build-esm": "tsc -p tsconfig.build.esm.json",
        "api-extractor": "mkdir -p api && [ -z \"${CI}\" ] && (api-extractor run -l) || (api-extractor run)",
        "test": "jest --watch",
        "test-once": "jest",
        "test-ci": "JEST_JUNIT_OUTPUT=./ci/results/test-results.xml jest --ci --config jest.ci.js",
        "eslint": "eslint $npm_package_config_eslint",
        "eslint-ci": "mkdir -p ./ci/results && eslint -f checkstyle -o ci/results/eslint-results.xml $npm_package_config_eslint",
        "prettier-check": "prettier --check '{src,test}/**/*.{ts,tsx,json,scss,md,yaml,html}'",
        "prettier-write": "prettier --write '{src,test}/**/*.{ts,tsx,json,scss,md,yaml,html}'",
        "dep-cruiser": "depcruise --validate .dependency-cruiser.js --output-type err-long src/",
        "dep-cruiser-ci": "depcruise --validate .dependency-cruiser.js --output-type err-long src/",
        "validate": "npm run dep-cruiser && npm run eslint && npm run prettier-check",
        "validate-ci": "npm run dep-cruiser-ci && npm run eslint-ci && npm run prettier-check"
    },
    "dependencies": {
        "@gooddata/sdk-backend-base": "^8.6.0-alpha.41",
        "@gooddata/sdk-backend-spi": "^8.6.0-alpha.41",
        "@gooddata/sdk-model": "^8.6.0-alpha.41",
        "lodash": "^4.17.19",
        "ts-invariant": "^0.7.3",
        "tslib": "^2.0.0"
    },
    "devDependencies": {
        "@gooddata/api-model-bear": "^8.6.0-alpha.41",
        "@gooddata/eslint-config": "^2.1.0",
        "@gooddata/reference-workspace": "^8.6.0-alpha.41",
        "@microsoft/api-documenter": "^7.12.13",
        "@microsoft/api-extractor": "^7.18.4",
        "@types/jest": "^26.0.12",
        "@types/lodash": "^4.14.158",
        "@typescript-eslint/eslint-plugin": "^4.14.2",
        "@typescript-eslint/parser": "^4.14.2",
        "concurrently": "^6.0.2",
        "dependency-cruiser": "^9.9.5",
        "eslint": "^7.5.0",
        "eslint-plugin-header": "^3.0.0",
        "eslint-plugin-import": "^2.22.0",
        "eslint-plugin-jest": "^24.1.3",
        "eslint-plugin-no-only-tests": "^2.4.0",
        "eslint-plugin-prettier": "^3.1.4",
        "eslint-plugin-sonarjs": "^0.9.1",
        "jest": "^26.4.2",
        "jest-junit": "^3.0.0",
        "prettier": "~2.3.0",
        "ts-jest": "^26.3.0",
        "typescript": "4.0.2"
    }
}
