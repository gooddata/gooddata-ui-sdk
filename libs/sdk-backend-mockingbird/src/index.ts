// (C) 2019-2022 GoodData Corporation
/**
 * This package provides a mock Analytical Backend implementation used mainly for testing.
 *
 * @remarks
 * For the GoodData platform version, see `@gooddata/sdk-backend-bear`.
 * For the GoodData.CN version, see `@gooddata/sdk-backend-tiger`.
 *
 * @packageDocumentation
 */
import { dummyBackend, dummyBackendEmptyData, dummyDataView } from "@gooddata/sdk-backend-base";
export { dummyBackend, dummyBackendEmptyData, dummyDataView };

export { recordedBackend, defaultRecordedBackendCapabilities } from "./recordedBackend";

export {
    RecordingIndex,
    ExecutionRecording,
    InsightRecording,
    DisplayFormRecording,
    ScenarioRecording,
    RecordedBackendConfig,
    CatalogRecording,
    VisClassesRecording,
    DashboardRecording,
    RecordedRefType,
    SecuritySettingsUrlValidator,
    SecuritySettingsOrganizationScope,
    SecuritySettingsPluginUrlValidator,
    IUserGroup,
    IUsers,
    IAccessControl,
    IUserManagement,
} from "./recordedBackend/types";

export {
    recordedDataView,
    recordedDataViews,
    recordedInsight,
    DataViewAll,
    dataViewWindow,
    DataViewFirstPage,
    NamedDataView,
} from "./recordedBackend/execution";

export { CompositeBackendPart, compositeBackend } from "./compositeBackend/index";

export {
    LegacyExecutionRecording,
    legacyRecordedBackend,
    legacyRecordedDataView,
    LegacyRecordingIndex,
    LegacyWorkspaceRecordings,
} from "./legacyRecordedBackend";

export {
    objRefsToStringKey,
    elementsQueryParamsFingerprint,
    elementsQueryParamsToElementsEntryId,
    elementsQueryParamsToRequestEntryId,
} from "./recordedBackend/utils";
