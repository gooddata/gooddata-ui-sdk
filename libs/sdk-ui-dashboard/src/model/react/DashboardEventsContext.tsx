// (C) 2019 GoodData Corporation
import React, { createContext, useContext, useMemo } from "react";
import noop from "lodash/noop.js";

import { DashboardEventHandler } from "../eventHandlers/eventHandler.js";

/**
 * @alpha
 */
export interface IDashboardEventsContext {
    registerHandler: (handler: DashboardEventHandler) => void;
    unregisterHandler: (handler: DashboardEventHandler) => void;
}

/**
 * @alpha
 */
const DashboardEventsContext = createContext<IDashboardEventsContext>({
    registerHandler: noop,
    unregisterHandler: noop,
});
DashboardEventsContext.displayName = "DashboardEventsContext";

/**
 * @alpha
 */
export const useDashboardEventsContext = (): IDashboardEventsContext => useContext(DashboardEventsContext);

/**
 * @internal
 */
export interface IDashboardEventsProvider extends IDashboardEventsContext {
    children: React.ReactNode;
}

/**
 * @internal
 */
export function DashboardEventsProvider(props: IDashboardEventsProvider): JSX.Element {
    const { children, registerHandler, unregisterHandler } = props;
    const value = useMemo(() => {
        return {
            registerHandler,
            unregisterHandler,
        };
    }, [registerHandler, unregisterHandler]);
    return <DashboardEventsContext.Provider value={value}>{children}</DashboardEventsContext.Provider>;
}
