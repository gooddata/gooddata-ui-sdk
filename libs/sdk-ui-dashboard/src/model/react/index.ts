// (C) 2021-2024 GoodData Corporation
export {
    DashboardStoreProvider,
    useDashboardDispatch,
    useDashboardSelector,
    ReactDashboardContext,
} from "./DashboardStoreProvider.js";
export type { IDashboardEventsContext } from "./DashboardEventsContext.js";
export { useDashboardEventsContext } from "./DashboardEventsContext.js";
export type { CommandProcessingStatus } from "./useDashboardCommandProcessing.js";
export { useDashboardCommandProcessing } from "./useDashboardCommandProcessing.js";
export { useDashboardEventDispatch } from "./useDashboardEventDispatch.js";
export type {
    QueryProcessingStatus,
    QueryProcessingErrorState,
    QueryProcessingPendingState,
    QueryProcessingRejectedState,
    QueryProcessingRunningState,
    QueryProcessingState,
    QueryProcessingSuccessState,
    UseDashboardQueryProcessingResult,
} from "./useDashboardQueryProcessing.js";
export { useDashboardQueryProcessing } from "./useDashboardQueryProcessing.js";
export { useDashboardUserInteraction } from "./useDashboardUserInteraction.js";
export type { UseDashboardAsyncRender } from "./useDashboardAsyncRender.js";
export { useDashboardAsyncRender } from "./useDashboardAsyncRender.js";
export type { IDashboardStoreProviderProps } from "./types.js";
export { useDispatchDashboardCommand } from "./useDispatchDashboardCommand.js";
export { useWidgetExecutionsHandler } from "./useWidgetExecutionsHandler.js";
export {
    useDashboardScheduledEmails,
    DEFAULT_MAX_AUTOMATIONS,
} from "./useDasboardScheduledEmails/useDashboardScheduledEmails.js";
export { useDashboardAlerts } from "./useDashboardAlerts.js";
export type { IUseWidgetSelectionResult } from "./useWidgetSelection.js";
export { useWidgetSelection } from "./useWidgetSelection.js";
export { useWidgetFilters } from "./useWidgetFilters.js";
export { useFiltersForDashboardScheduledExport } from "./useDasboardScheduledEmails/useFiltersForDashboardScheduledExport.js";
export type { IUseFiltersForDashboardScheduledExportProps } from "./useDasboardScheduledEmails/useFiltersForDashboardScheduledExport.js";
export { useFiltersForWidgetScheduledExport } from "./useDasboardScheduledEmails/useFiltersForWidgetScheduledExport.js";
export type { IUseFiltersForWidgetScheduledExportProps } from "./useDasboardScheduledEmails/useFiltersForWidgetScheduledExport.js";
export { useDashboardScheduledEmailsFilters } from "./useDasboardScheduledEmails/useDashboardScheduledEmailsFilters.js";
export type { IUseDashboardScheduledEmailsFiltersProps } from "./useDasboardScheduledEmails/useDashboardScheduledEmailsFilters.js";
export { useDashboardAutomations } from "./useDasboardScheduledEmails/useDashboardAutomations.js";
export { useWorkspaceUsers } from "./useWorkspaceUsers.js";
