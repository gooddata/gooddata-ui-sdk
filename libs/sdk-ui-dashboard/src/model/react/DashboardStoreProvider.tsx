// (C) 2021-2022 GoodData Corporation
import React from "react";
import { AnyAction, Dispatch } from "@reduxjs/toolkit";
import { createDispatchHook, createSelectorHook, Provider, TypedUseSelectorHook } from "react-redux";
import { DashboardState } from "../store/index.js";
import { DashboardEventsProvider } from "./DashboardEventsContext.js";
import { useInitializeDashboardStore } from "./useInitializeDashboardStore.js";
import { IDashboardStoreProviderProps } from "./types.js";

/**
 * @alpha
 */
export const ReactDashboardContext: any = React.createContext(null);

/**
 * @alpha
 */
export const useDashboardDispatch: () => Dispatch<AnyAction> = createDispatchHook(ReactDashboardContext);

/**
 * Hook for retrieving data from the dashboard state.
 *
 * @example
 * Example how to obtain all insights stored on the dashboard:
 *
 * ```tsx
 * import { useDashboardSelector, selectInsights } from "@gooddata/sdk-ui-dashboard";
 *
 * const CustomDashboardWidget = () => {
 *   const insights = useDashboardSelector(selectInsights);
 *
 *   return (
 *      <pre>{JSON.stringify(insights, null, 2)}</pre>
 *   );
 * }
 * ```
 *
 * @public
 */
export const useDashboardSelector: TypedUseSelectorHook<DashboardState> =
    createSelectorHook(ReactDashboardContext);

/**
 * @internal
 */
export const DashboardStoreProvider: React.FC<IDashboardStoreProviderProps> = (props) => {
    const dashboardStore = useInitializeDashboardStore(props);
    const { additionalReduxContext } = props;

    if (!dashboardStore) {
        return null;
    }

    if (additionalReduxContext != undefined && additionalReduxContext !== ReactDashboardContext) {
        /*
         * Setting store into multiple contexts is essential in environments where a dynamically loaded
         * dashboard engine is integrated into an application that adds additional embedded, local plugins
         * which use Redux hooks (useDashboardSelector, useDashboardDispatch)
         *
         * Such local code is typically built against a local version of the sdk-ui-dashboard and has
         * different redux context providers from those used by dynamically loaded engine bundle.
         *
         * When the local code uses redux hooks, it will explode because they look into a different context
         * where the store is unset.
         */
        return (
            <Provider store={dashboardStore.store} context={ReactDashboardContext}>
                <Provider store={dashboardStore.store} context={additionalReduxContext}>
                    <DashboardEventsProvider
                        registerHandler={dashboardStore.registerEventHandler}
                        unregisterHandler={dashboardStore.unregisterEventHandler}
                    >
                        {props.children}
                    </DashboardEventsProvider>
                </Provider>
            </Provider>
        );
    }

    return (
        <Provider store={dashboardStore.store} context={ReactDashboardContext}>
            <DashboardEventsProvider
                registerHandler={dashboardStore.registerEventHandler}
                unregisterHandler={dashboardStore.unregisterEventHandler}
            >
                {props.children}
            </DashboardEventsProvider>
        </Provider>
    );
};
