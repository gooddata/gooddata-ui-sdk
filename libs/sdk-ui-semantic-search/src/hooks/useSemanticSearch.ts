// (C) 2024 GoodData Corporation

import { useState, useEffect } from "react";
import {
    ISemanticSearchResultItem,
    GenAISemanticSearchType,
    ISemanticSearchRelationship,
} from "@gooddata/sdk-model";
import { useBackendStrict, useWorkspaceStrict } from "@gooddata/sdk-ui";
import { IAnalyticalBackend } from "@gooddata/sdk-backend-spi";

/**
 * The result of the semantic search hook.
 * @beta
 */
export type SemanticSearchInputResult = {
    /**
     * Flag indicating whether the search is in progress.
     * - idle - means we did not search yet
     * - loading - means we are currently searching
     * - error - means the search failed
     * - success - means the search succeeded
     */
    searchStatus: "idle" | "loading" | "error" | "success";
    /**
     * Error message if the search failed.
     */
    searchError: string;
    /**
     * The search results.
     */
    searchResults: ISemanticSearchResultItem[];
    relationships: ISemanticSearchRelationship[];
};

/**
 * Default state of the semantic search hook.
 */
const DEFAULT_STATE: SemanticSearchInputResult = {
    searchStatus: "idle",
    searchError: "",
    searchResults: [],
    relationships: [],
};

/**
 * Input parameters for the semantic search hook.
 * @beta
 */
export type SemanticSearchHookInput = {
    /**
     * The search term.
     */
    searchTerm: string;
    /**
     * The object types to search for.
     */
    objectTypes?: GenAISemanticSearchType[];
    /**
     * Whether to perform deep search.
     * E.g. find dashboard if it contains matching insight.
     */
    deepSearch?: boolean;
    /**
     * The number of results to return.
     */
    limit?: number;
    /**
     * The backend to use for the search.
     * If omitted, will be retrieved from the context.
     */
    backend?: IAnalyticalBackend;
    /**
     * The workspace to use for the search.
     * If omitted, will be retrieved from the context.
     */
    workspace?: string;
};

const DEFAULT_OBJECT_TYPES: GenAISemanticSearchType[] = [];

/**
 * Hook to perform semantic search.
 * Makes the request to server and returns the search results.
 * The request is cancellable and auto-updates when the search term changes.
 * @beta
 */
export const useSemanticSearch = ({
    searchTerm,
    objectTypes = DEFAULT_OBJECT_TYPES,
    deepSearch,
    limit,
    backend,
    workspace,
}: SemanticSearchHookInput): SemanticSearchInputResult => {
    const [state, setState] = useState<SemanticSearchInputResult>(DEFAULT_STATE);
    const effectiveBackend = useBackendStrict(backend);
    const effectiveWorkspace = useWorkspaceStrict(workspace);

    useEffect(() => {
        if (searchTerm === "") {
            setState(DEFAULT_STATE);
            return;
        }

        setState((oldState) => ({
            // Keep the previous state of results/error, only update the loading flag
            ...oldState,
            searchStatus: "loading",
        }));

        // Construct the query
        let qb = effectiveBackend
            .workspace(effectiveWorkspace)
            .genAI()
            .getSemanticSearchQuery()
            .withQuestion(searchTerm);
        if (objectTypes.length) {
            qb = qb.withObjectTypes(objectTypes);
        }
        if (deepSearch) {
            qb = qb.withDeepSearch(deepSearch);
        }
        if (limit) {
            qb = qb.withLimit(limit);
        }

        // Execute the query
        const searchController = new AbortController();
        qb.query({
            signal: searchController.signal,
        })
            .then((res) => {
                setState({
                    searchStatus: "success",
                    searchError: "",
                    searchResults: res.results,
                    relationships: res.relationships,
                });
            })
            .catch((e) => {
                if (searchController.signal.aborted)
                    // Ignore Aborted error
                    return;

                setState({
                    searchStatus: "error",
                    searchError: errorMessage(e),
                    searchResults: [],
                    relationships: [],
                });
            });

        return () => {
            searchController.abort();
        };
    }, [effectiveBackend, effectiveWorkspace, searchTerm, objectTypes, deepSearch, limit]);

    return state;
};

/**
 * Retrieve error message from the error object.
 */
const errorMessage = (e: unknown) => {
    if (e instanceof Error) {
        return e.message;
    }
    return String(e);
};
