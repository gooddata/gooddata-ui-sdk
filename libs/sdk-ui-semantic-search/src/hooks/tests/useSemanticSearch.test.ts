// (C) 2024 GoodData Corporation

import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import { act, renderHook } from "@testing-library/react";
import { dummyBackend } from "@gooddata/sdk-backend-mockingbird";
import { useSemanticSearch } from "../useSemanticSearch.js";

const backend = dummyBackend();

describe("useSemanticSearch hook", () => {
    beforeEach(() => {
        vi.useFakeTimers();
    });

    afterEach(() => {
        vi.useRealTimers();
    });

    it("should return correct results", async () => {
        const { result } = renderHook(() =>
            useSemanticSearch({
                backend,
                workspace: "test",
                searchTerm: "test",
            }),
        );

        expect(result.current).toEqual({
            searchStatus: "loading",
            searchResults: [],
            searchError: "",
            relationships: [],
        });

        await act(() => vi.advanceTimersByTimeAsync(150));

        expect(result.current.searchStatus).toEqual("success");
        expect(result.current).toMatchSnapshot();
    });

    it("should cancel previous search", async () => {
        // This test relies on dummy backend delay in returning results in 100ms
        const DELAY = 100;
        const { result, rerender } = renderHook(
            ({ searchTerm }) =>
                useSemanticSearch({
                    backend,
                    workspace: "test",
                    searchTerm,
                }),
            {
                initialProps: { searchTerm: "test" },
            },
        );

        expect(result.current).toEqual({
            searchStatus: "loading",
            searchResults: [],
            searchError: "",
            relationships: [],
        });

        // Simulate user entering a different search query before the first request is completed
        await act(() => vi.advanceTimersByTimeAsync(DELAY * 0.75));
        rerender({ searchTerm: "test2" });

        // Now, if the first request is not cancelled, the search results should have been delivered by now
        await act(() => vi.advanceTimersByTimeAsync(DELAY * 0.5));

        expect(result.current).toEqual({
            searchStatus: "loading",
            searchResults: [],
            searchError: "",
            relationships: [],
        });

        // Let the second request complete
        await act(() => vi.advanceTimersByTimeAsync(DELAY * 0.5));

        expect(result.current.searchStatus).toEqual("success");
        expect(result.current).toMatchSnapshot();
    });
});
