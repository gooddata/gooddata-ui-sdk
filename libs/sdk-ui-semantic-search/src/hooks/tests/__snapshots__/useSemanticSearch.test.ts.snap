// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useSemanticSearch hook > should cancel previous search 1`] = `
{
  "relationships": [],
  "searchError": "",
  "searchResults": [
    {
      "createdAt": "2023-08-03T13:17:26.923537",
      "description": "test2",
      "id": "dataset",
      "modifiedAt": "2023-08-03T13:17:26.923537",
      "score": 0.5,
      "scoreDescriptor": 0.5,
      "scoreExactMatch": 0,
      "scoreTitle": 0.5,
      "tags": [],
      "title": "dataset title",
      "type": "dataset",
      "visualizationUrl": undefined,
      "workspaceId": "test",
    },
    {
      "createdAt": "2023-08-03T13:17:26.923537",
      "description": "test2",
      "id": "attribute",
      "modifiedAt": "2023-08-03T13:17:26.923537",
      "score": 0.5,
      "scoreDescriptor": 0.5,
      "scoreExactMatch": 0,
      "scoreTitle": 0.5,
      "tags": [],
      "title": "attribute title",
      "type": "attribute",
      "visualizationUrl": undefined,
      "workspaceId": "test",
    },
    {
      "createdAt": "2023-08-03T13:17:26.923537",
      "description": "test2",
      "id": "label",
      "modifiedAt": "2023-08-03T13:17:26.923537",
      "score": 0.5,
      "scoreDescriptor": 0.5,
      "scoreExactMatch": 0,
      "scoreTitle": 0.5,
      "tags": [],
      "title": "label title",
      "type": "label",
      "visualizationUrl": undefined,
      "workspaceId": "test",
    },
    {
      "createdAt": "2023-08-03T13:17:26.923537",
      "description": "test2",
      "id": "fact",
      "modifiedAt": "2023-08-03T13:17:26.923537",
      "score": 0.5,
      "scoreDescriptor": 0.5,
      "scoreExactMatch": 0,
      "scoreTitle": 0.5,
      "tags": [],
      "title": "fact title",
      "type": "fact",
      "visualizationUrl": undefined,
      "workspaceId": "test",
    },
    {
      "createdAt": "2023-08-03T13:17:26.923537",
      "description": "test2",
      "id": "date",
      "modifiedAt": "2023-08-03T13:17:26.923537",
      "score": 0.5,
      "scoreDescriptor": 0.5,
      "scoreExactMatch": 0,
      "scoreTitle": 0.5,
      "tags": [],
      "title": "date title",
      "type": "date",
      "visualizationUrl": undefined,
      "workspaceId": "test",
    },
    {
      "createdAt": "2023-08-03T13:17:26.923537",
      "description": "test2",
      "id": "metric",
      "modifiedAt": "2023-08-03T13:17:26.923537",
      "score": 0.5,
      "scoreDescriptor": 0.5,
      "scoreExactMatch": 0,
      "scoreTitle": 0.5,
      "tags": [],
      "title": "metric title",
      "type": "metric",
      "visualizationUrl": undefined,
      "workspaceId": "test",
    },
    {
      "createdAt": "2023-08-03T13:17:26.923537",
      "description": "test2",
      "id": "visualization",
      "modifiedAt": "2023-08-03T13:17:26.923537",
      "score": 0.5,
      "scoreDescriptor": 0.5,
      "scoreExactMatch": 0,
      "scoreTitle": 0.5,
      "tags": [],
      "title": "visualization title",
      "type": "visualization",
      "visualizationUrl": "local:line",
      "workspaceId": "test",
    },
    {
      "createdAt": "2023-08-03T13:17:26.923537",
      "description": "test2",
      "id": "dashboard",
      "modifiedAt": "2023-08-03T13:17:26.923537",
      "score": 0.5,
      "scoreDescriptor": 0.5,
      "scoreExactMatch": 0,
      "scoreTitle": 0.5,
      "tags": [],
      "title": "dashboard title",
      "type": "dashboard",
      "visualizationUrl": undefined,
      "workspaceId": "test",
    },
  ],
  "searchStatus": "success",
}
`;

exports[`useSemanticSearch hook > should return correct results 1`] = `
{
  "relationships": [],
  "searchError": "",
  "searchResults": [
    {
      "createdAt": "2023-08-03T13:17:26.923537",
      "description": "test",
      "id": "dataset",
      "modifiedAt": "2023-08-03T13:17:26.923537",
      "score": 0.5,
      "scoreDescriptor": 0.5,
      "scoreExactMatch": 0,
      "scoreTitle": 0.5,
      "tags": [],
      "title": "dataset title",
      "type": "dataset",
      "visualizationUrl": undefined,
      "workspaceId": "test",
    },
    {
      "createdAt": "2023-08-03T13:17:26.923537",
      "description": "test",
      "id": "attribute",
      "modifiedAt": "2023-08-03T13:17:26.923537",
      "score": 0.5,
      "scoreDescriptor": 0.5,
      "scoreExactMatch": 0,
      "scoreTitle": 0.5,
      "tags": [],
      "title": "attribute title",
      "type": "attribute",
      "visualizationUrl": undefined,
      "workspaceId": "test",
    },
    {
      "createdAt": "2023-08-03T13:17:26.923537",
      "description": "test",
      "id": "label",
      "modifiedAt": "2023-08-03T13:17:26.923537",
      "score": 0.5,
      "scoreDescriptor": 0.5,
      "scoreExactMatch": 0,
      "scoreTitle": 0.5,
      "tags": [],
      "title": "label title",
      "type": "label",
      "visualizationUrl": undefined,
      "workspaceId": "test",
    },
    {
      "createdAt": "2023-08-03T13:17:26.923537",
      "description": "test",
      "id": "fact",
      "modifiedAt": "2023-08-03T13:17:26.923537",
      "score": 0.5,
      "scoreDescriptor": 0.5,
      "scoreExactMatch": 0,
      "scoreTitle": 0.5,
      "tags": [],
      "title": "fact title",
      "type": "fact",
      "visualizationUrl": undefined,
      "workspaceId": "test",
    },
    {
      "createdAt": "2023-08-03T13:17:26.923537",
      "description": "test",
      "id": "date",
      "modifiedAt": "2023-08-03T13:17:26.923537",
      "score": 0.5,
      "scoreDescriptor": 0.5,
      "scoreExactMatch": 0,
      "scoreTitle": 0.5,
      "tags": [],
      "title": "date title",
      "type": "date",
      "visualizationUrl": undefined,
      "workspaceId": "test",
    },
    {
      "createdAt": "2023-08-03T13:17:26.923537",
      "description": "test",
      "id": "metric",
      "modifiedAt": "2023-08-03T13:17:26.923537",
      "score": 0.5,
      "scoreDescriptor": 0.5,
      "scoreExactMatch": 0,
      "scoreTitle": 0.5,
      "tags": [],
      "title": "metric title",
      "type": "metric",
      "visualizationUrl": undefined,
      "workspaceId": "test",
    },
    {
      "createdAt": "2023-08-03T13:17:26.923537",
      "description": "test",
      "id": "visualization",
      "modifiedAt": "2023-08-03T13:17:26.923537",
      "score": 0.5,
      "scoreDescriptor": 0.5,
      "scoreExactMatch": 0,
      "scoreTitle": 0.5,
      "tags": [],
      "title": "visualization title",
      "type": "visualization",
      "visualizationUrl": "local:line",
      "workspaceId": "test",
    },
    {
      "createdAt": "2023-08-03T13:17:26.923537",
      "description": "test",
      "id": "dashboard",
      "modifiedAt": "2023-08-03T13:17:26.923537",
      "score": 0.5,
      "scoreDescriptor": 0.5,
      "scoreExactMatch": 0,
      "scoreTitle": 0.5,
      "tags": [],
      "title": "dashboard title",
      "type": "dashboard",
      "visualizationUrl": undefined,
      "workspaceId": "test",
    },
  ],
  "searchStatus": "success",
}
`;
